### PR TITLE
drivers: add API for drivers early initialization

### PR DESCRIPTION
**Pull request by XuNeo**
## Summary

Add `drivers_early_initialize` for drivers not reply on system resource. `note_sysview_initialize` is one of them to allow note to work as early as possible.

## Impact
No.

## Testing

stm32f429i-disco with systemview.

### CODE REVIEW

File: sched/init/nx_start.c [hunk 1]
```diff
@@ -520,6 +520,8 @@ void nx_start(void)
 
   /* Initialize RTOS Data ***************************************************/
 
+  drivers_early_initialize();
```
Comment (anchao) [inline sched/init/nx_start.c:523]: I'm curious how you ensure that all early initialize can be executed before idle tcb initialized.

Comment (XuNeo) [reply sched/init/nx_start.c:523]: I'm not sure if I understand correctly. Currently `drivers_early_initialize` calls `notesnap_register` etc. which do not rely on idle tcb intialization. What's you concern here?

Comment (anchao) [reply sched/init/nx_start.c:523]: Are you sure you have run the current trace function? There is a potential issue here

```
nx_start
|
 ->sched_trace_begin
   |
    ->trace_begin
      |
       ->sched_note_begin
         |
          ->sched_note_event
            |
             ->sched_note_event_ip
               |
                ->this_task()
```
                
this_task will be NULL and unable to record the tcb pid, cpu, and priority correctly. how abort move `sched_trace_begin() ` after idle initialization?

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by acassis** [comment]
@XuNeo please include Documentation/ about it. Please create some entry here: https://nuttx.apache.org/docs/latest/components/drivers/index.html explaining about early driver initialization.

I think NuttX is very confuse about it driver initialization: you can do it on NSH Arch init; Board Initialization; Early Initialization; Late Initialization; Board CTL and now you added another option! Hehehe

@patacongo  explained about some of those difference 10 years ago:
https://acassis.wordpress.com/2014/09/20/how-to-do-late-board-initialization-on-nuttx/

But I think there are more complex now. @xiaoxiang781216 did you get the guy that was suppose to work on NuttX Documentation? Maybe he/she could help us on that.

**Comment by acassis** [comment]
Hmm, maybe we can create a DOC from my other blog post:

https://acassis.wordpress.com/2015/07/01/nuttx-board-initialization-where-can-i-put-my-code-to-be-execute/

File: sched/init/nx_start.c [hunk 2]
```diff
@@ -534,6 +530,12 @@ void nx_start(void)
 
   g_nx_initstate = OSINIT_TASKLISTS;
 
+  /* Initialize RTOS Data ***************************************************/
+
+  drivers_early_initialize();
```
Comment (anchao) [inline sched/init/nx_start.c:535]: I still feel a little strange about **drivers_early_initialize()**. Driver early initialization is executed before core resource initialization. Does that mean related early initialization such as uart(*****_earlyserialinit()**) could be placed in this implementation?

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by patacongo** [comment]
> @patacongo explained about some of those difference 10 years ago:
> https://acassis.wordpress.com/2014/09/20/how-to-do-late-board-initialization-on-nuttx/

You might be talking about my initialization documentation here:  https://cwiki.apache.org/confluence/display/NUTTX/NuttX+Initialization+Sequence

**Comment by acassis** [comment]
> > @patacongo explained about some of those difference 10 years ago:
> > https://acassis.wordpress.com/2014/09/20/how-to-do-late-board-initialization-on-nuttx/
> 
> You might be talking about my initialization documentation here: https://cwiki.apache.org/confluence/display/NUTTX/NuttX+Initialization+Sequence

Yes, we need first move it to our official documentation site, that is linked in initial page: https://nuttx.apache.org/

BTW, I will open a public discussion at mailing list about it: how to simplify NuttX initialization?

File: sched/init/nx_start.c (hunk 2 above)
Comment (XuNeo) [reply sched/init/nx_start.c:535]: Compared to existing uart(***_earlyserialinit()), they are similar but uart needs to init earlier.

As you have commented out, the sched note facility does need idle tcb to be ready at least.

The bringup process is too compilicated for me to imagine the perfect solution for my coming issue that segger systemview cannot work earlier than memory heap initialization, making it unable to detect correct heap information. Thus this patch comes.

Feel free to make any adjustment to my code as neccesary.

Comment (anchao) [reply sched/init/nx_start.c:535]: I don't have better solution with current implementation, let's keep moving

**Pull request merged**